cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -105,6 +119,103 @@ protected void shutDown() throws Exception {
+          spark, spark, SchedulableProgramType.SPARK));
+        break;
+      case CUSTOM_ACTION:
+        actionSpec = workflowSpec.getCustomActionMap().get(actionInfo.getProgramName());
+        break;
+      default:
+        LOG.error("Unknown Program Type '{}', Program '{}' in the Workflow.", actionInfo.getProgramType(),
+                  actionInfo.getProgramName());
+        throw new IllegalStateException("Workflow stopped without executing all tasks");
+    }
+
+    WorkflowStatus workflowStatus = new WorkflowStatus(state(), actionSpec, node.getNodeId());
+    status.put(node.getNodeId(), workflowStatus);
+
+    WorkflowAction action = initialize(actionSpec, classLoader, instantiator);
+    try {
+      ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(action.getClass().getClassLoader());
+      try {
+        action.run();
+      } finally {
+        ClassLoaders.setContextClassLoader(oldClassLoader);
+      }
+    } catch (Throwable t) {
+      LOG.warn("Exception on WorkflowAction.run(), aborting Workflow. {}", actionSpec);
+      // this will always rethrow
+      Throwables.propagateIfPossible(t, Exception.class);
+    } finally {
+      // Destroy the action.
+      destroy(actionSpec, action);
+    }
+    status.remove(node.getNodeId());
+  }
+
+  private void executeNode(final ApplicationSpecification appSpec, WorkflowNode node,
+                           final InstantiatorFactory instantiator, final ClassLoader classLoader) throws Exception {
+    WorkflowNodeType nodeType = node.getType();
+    switch (nodeType) {
+      case ACTION:
+        executeWorkflowAction(appSpec, (WorkflowActionNode) node, instantiator, classLoader);
+        break;
+      case FORK:
+        WorkflowForkNode fork = (WorkflowForkNode) node;
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        CompletionService<String> completionService =
+          new ExecutorCompletionService<String>(executorService);
+
+        for (final WorkflowForkBranch branch : fork.getBranches()) {
+          completionService.submit(new Callable<String>() {
+            @Override
+            public String call() throws Exception {
+              Iterator<WorkflowNode> iterator = branch.getNodes().iterator();
+              while (!Thread.currentThread().isInterrupted() && iterator.hasNext() && running) {
+                executeNode(appSpec, iterator.next(), instantiator, classLoader);
+              }
+              return branch.toString();
+            }
+          });
+        }
+
+        try {
+          for (int i = 0; i < fork.getBranches().size(); i++) {
+            Future<String> f = completionService.take();
+            try {
+              LOG.info("Execution of branch {} for fork {} completed", f.get(), fork);
+            } catch (ExecutionException ex) {
+              LOG.info("Exception occurred in the execution of the fork node {}", fork);
+              throw ex;
```

Comment:
Ideally you should cancel the Futures of all the submitted jobs so that in the implementation you can use it as a signal to kill the job (e.g. call MapReduceProgramController.stop()). Otherwise you might still need to wait for a long time even you know the Workflow is going to be failed.